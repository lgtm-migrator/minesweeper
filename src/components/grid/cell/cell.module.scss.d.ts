// autogenerated by typings-for-css-modules-loader.
// Please do not change this file!
declare namespace CellModuleScssNamespace {
  export interface ICellModuleScss {
    bombEntity: string;
    bombFrame: string;
    changedBorderColor: string;
    closedFrame: string;
    emptyFrame: string;
    transparentClosedFrame: string;
  }
}

declare const CellModuleScssModule: CellModuleScssNamespace.ICellModuleScss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: CellModuleScssNamespace.ICellModuleScss;
};

export = CellModuleScssModule;
