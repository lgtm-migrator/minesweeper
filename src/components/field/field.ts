import { resolve } from "path";

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coordinates = [number, number];

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  mark: 11,
  weakMark: 12,
};

export const generateFieldWithDefaultState = (size: number, state: Cell = CellState.empty): Field =>
  new Array(size).fill(null).map(() => new Array(size).fill(state));

export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error("Probability must be between 0 and 1");
  }

  const result: Field = generateFieldWithDefaultState(size);

  // the size grid of the cells
  let unprocessedCells = size * size;
  // cells with bombs included
  let restCellsWithBombs = unprocessedCells * probability;

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (restCellsWithBombs === 0) {
        return result;
      }
      if (restCellsWithBombs / unprocessedCells > 0) {
        result[y][x] = CellState.bomb;
        restCellsWithBombs--;
      }
      unprocessedCells--;
    }
  }

  return result;
};
