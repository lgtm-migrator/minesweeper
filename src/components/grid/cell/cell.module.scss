.closedFrame {
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;

  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;

  &:hover {
    filter: brightness(1.1);
  }
}

.changedBorderColor {
  border-color: white #9e9e9e #9e9e9e white;
}

.transparentClosedFrame {
  border-color: transparent;
}

.revealedFrame {
  border-color: #dddddd;
  cursor: default;

  &:hover {
    filter: brightness(1);
  }
}

.bombFrame {
  background-color: #ec433c;
}

.bombEntity {
  border-radius: 50%;
  width: 1vh;
  height: 1vh;
  background-color: #333;
}

.flagComponent {
  width: 0px;
  height: 0px;

  border-top: 0.5vw solid transparent;
  border-bottom: 0.5vw solid transparent;
  border-left: 0.5vw solid #ec433c;
}

.flagTransparent {
  border-left: 0.5vw solid #f19996;
}
