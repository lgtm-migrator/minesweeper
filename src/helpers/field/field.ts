import { incrementNeighbours } from "../cells";

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coordinates = [number, number];

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  mark: 11,
  weakMark: 12,
};

/**
 * Create empty field
 *
 * Example generateFieldWithDefaultState(2)
 * [
 *   [0, 0],
 *   [0, 0],
 * ]
 *
 * @param {number} size
 * @param {Cell} state
 * @returns {Field}
 */
export const generateFieldWithDefaultState = (size: number, state: Cell = CellState.empty): Field =>
  new Array(size).fill(null).map(() => new Array(size).fill(state));

/**
 * Generate field with mines
 * Cell has a range between 0 and 9 where
 * 0 - have no any mine
 * from 0 to 8 - count of mines arrount the cell
 * 9 - special flag for the mine
 *
 * (1/9 ~ 11.11%)
 * Example fieldGenerator(3, 1/9) would return
 * [
 *   [1, 1, 1],
 *   [1, 9, 1],
 *   [1, 1, 1]
 * ]
 *
 * @param {number} size
 * @param {number} probability
 * @return {Field}
 */
export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error("Probability must be between 0 and 1");
  }

  const result: Field = generateFieldWithDefaultState(size);

  // the size grid of the cells
  let unprocessedCells = size * size;
  // cells filled with bombs
  let restCellsWithBombs = unprocessedCells * probability;

  // Stryker disable next-line EqualityOperator
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      // Stryker disable next-line EqualityOperator
      if (restCellsWithBombs / unprocessedCells > Math.random()) {
        result[y][x] = CellState.bomb;
        incrementNeighbours([y, x], result);
        restCellsWithBombs--;
      }
      unprocessedCells--;
    }
  }

  return result;
};
