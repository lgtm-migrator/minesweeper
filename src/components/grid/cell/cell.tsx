import { CellState, Cell as CellType, Coordinates } from "@helpers/field";
import { useMouseDown } from "~hooks/useMouseDown";
import classNames from "classnames";
import React, { FunctionComponent, PropsWithChildren, ReactNode } from "react";

import styles from "./cell.module.scss";

export interface CellProps {
  /**
   * Cell status based on the CellType
   */
  children: CellType;
  /**
   * Cell coordinates
   */
  coords: Coordinates;
  /**
   * onClick by cell handler
   */
  onClick: (coords: Coordinates) => void;
  /**
   * onContextMenu by cell handler
   */
  onContextMenu: (coords: Coordinates) => void;
}

export interface ComponentsMapProps {
  children?: CellType;
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
  onContextMenu?: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown?: () => void;
  onMouseUp?: () => void;
  onMouseLeave?: () => void;
  mouseDown?: boolean;
  "data-testid"?: string;
  role?: string;
}

type RevealedProps = Pick<ComponentsMapProps, "onContextMenu" | "data-testid" | "role">;

type ExcludeMouseDown = Omit<ComponentsMapProps, "mouseDown">;

export const isActiveCell = (cell: CellType): boolean =>
  [CellState.hidden, CellState.mark, CellState.weakMark].includes(cell);

export const areEqual = (prevProps: CellProps, nextProps: CellProps): boolean => {
  const areEqualCoords = prevProps.coords.filter((coord, index) => nextProps.coords[index] !== coord).length === 0;

  return (
    prevProps.children === nextProps.children &&
    areEqualCoords &&
    prevProps.onClick === nextProps.onClick &&
    prevProps.onContextMenu === nextProps.onContextMenu
  );
};

export const CellComponent: FunctionComponent<CellProps> = ({ coords, children, ...restProps }) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  const onClick = () => restProps.onClick(coords);

  const onContextMenu = (event: React.MouseEvent<HTMLDivElement>) => {
    /**
     * Prevent context menu by default
     */
    event.preventDefault();

    if (isActiveCell(children)) {
      restProps.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    mouseDown,
    "data-testid": `${coords}`,
    role: "cell",
  };

  return <ComponentsMap {...props}>{children}</ComponentsMap>;
};

export const ComponentsMap: FunctionComponent<ComponentsMapProps> = ({ children, ...rest }) => {
  const nonActiveCellProps: RevealedProps = {
    onContextMenu: rest.onContextMenu,
    "data-testid": rest["data-testid"],
    role: rest.role,
  };

  switch (children) {
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <div className={styles.bombEntity} data-testid={`bomb_${rest["data-testid"]}`} />
        </BombFrame>
      );
    case CellState.mark:
      return (
        <ClosedFrame {...rest}>
          <FlagComponent data-testid={`flag_${rest["data-testid"]}`} />
        </ClosedFrame>
      );

    case CellState.weakMark:
      return (
        <ClosedFrame {...rest}>
          <WeakFlagComponent data-testid={`weakFlag_${rest["data-testid"]}`} />
        </ClosedFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...rest} />;
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>;
  }
};

export const ClosedFrame: FunctionComponent<PropsWithChildren<{ mouseDown?: boolean }>> = ({
  mouseDown,
  children,
  ...restProps
}) => {
  return (
    <div
      {...(restProps as ExcludeMouseDown)}
      className={classNames(styles.closedFrame, {
        [styles.changedBorderColor]: !mouseDown,
        [styles.transparentClosedFrame]: mouseDown,
      })}
    >
      {children ?? null}
    </div>
  );
};

export const RevealedFrame: FunctionComponent<PropsWithChildren<RevealedProps>> = ({ children, ...restProps }) => {
  return (
    <div
      {...(restProps as ExcludeMouseDown)}
      className={classNames(styles.closedFrame, styles.revealedFrame)}
      style={{
        color: colors[children as CellType] ?? transparent,
      }}
    >
      {children}
    </div>
  );
};

export const BombFrame: FunctionComponent<{ children: ReactNode }> = ({ children, ...restProps }) => {
  return (
    <div
      {...(restProps as ExcludeMouseDown)}
      className={classNames(styles.closedFrame, styles.revealedFrame, styles.bombFrame)}
    >
      {children}
    </div>
  );
};

const FlagComponent: FunctionComponent = ({ ...restProps }) => {
  return <div {...(restProps as ExcludeMouseDown)} className={styles.flagComponent} />;
};

const WeakFlagComponent: FunctionComponent = ({ ...restProps }) => {
  return (
    <div {...(restProps as ExcludeMouseDown)} className={classNames(styles.flagComponent, styles.flagTransparent)} />
  );
};

const transparent = "rgba(0,0,0,0)";
const colors: { [key in CellType]: string } = {
  0: transparent,
  1: "#2a48ec",
  2: "#2bb13d",
  3: "#ec6561",
  4: "#233db7",
  5: "#a6070f",
  6: "#e400af",
  7: "#906a02",
  8: "#fa0707",
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};
