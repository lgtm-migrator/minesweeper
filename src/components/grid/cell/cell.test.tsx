import { CellState, Coordinates } from "@helpers/field";
import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";

import { CellComponent } from "./cell";

describe("Testing the CELL component", function () {
  const coords: Coordinates = [1, 1];

  for (let cell = CellState.empty; cell <= CellState.weakMark; cell++) {
    it("should check preventDefault contextMenu for every type of cell", function () {
      const internalProps = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<CellComponent children={cell} {...internalProps} />);

      expect(true).toBe(true);
    });

    it("onClick and onContextMenu handler should be called for active cells", function () {
      expect(true).toBe(true);
    });
  }
});
