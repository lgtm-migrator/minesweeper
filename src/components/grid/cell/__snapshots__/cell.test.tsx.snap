// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the CELL component Cell renders correct 1`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgba(0, 0, 0, 0);"
  >
    0
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 2`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(42, 72, 236);"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 3`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(43, 177, 61);"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 4`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(236, 101, 97);"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 5`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(35, 61, 183);"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 6`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(166, 7, 15);"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 7`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(228, 0, 175);"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 8`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(144, 106, 2);"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 9`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame"
    data-testid="1,1"
    role="cell"
    style="color: rgb(250, 7, 7);"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 10`] = `
<DocumentFragment>
  <div
    class="closedFrame revealedFrame bombFrame"
    data-testid="1,1"
    role="cell"
  >
    <div
      class="bombEntity"
      data-testid="bomb_1,1"
    />
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 11`] = `
<DocumentFragment>
  <div
    class="closedFrame changedBorderColor"
    data-testid="1,1"
    role="cell"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 12`] = `
<DocumentFragment>
  <div
    class="closedFrame changedBorderColor"
    data-testid="1,1"
    role="cell"
  >
    <div
      class="flagComponent"
      data-testid="flag_1,1"
    />
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Cell renders correct 13`] = `
<DocumentFragment>
  <div
    class="closedFrame changedBorderColor"
    data-testid="1,1"
    role="cell"
  >
    <div
      class="flagComponent flagTransparent"
      data-testid="weakFlag_1,1"
    />
  </div>
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 1`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 2`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 3`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 4`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 5`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 6`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 7`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 8`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 9`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 10`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 11`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 12`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;

exports[`Testing the CELL component Closed Frame renders correct 13`] = `
<DocumentFragment>
  <div
    class="closedFrame transparentClosedFrame"
  />
</DocumentFragment>
`;
